### PR TITLE
refactor: Reorganize and document `pytorch_trainer`

### PR DESCRIPTION
**Pull request by robinholzi**
# Motivation

I wanted to find out how different performance measurements were implemented in the pytorch trainer to be able to reason about them. I found it hard to follow the control flow though so I restructured the file a bit.

# Changes

What I did:
- Changed function order (moved core training loop function to the top and internal helper methods to the end.
- Added comment dividers
- Created a small control flow mermaid state diagram so new people can get a grasp of the training control flow more quickly.
- adjusted the names of local function names for consistency (made some private, and more important ones public)

# Review notice

The diff looks massive, however, the changes here are not really worth reviewing, as I didn't touch the function bodies except from the few cases where I left a gh-comment.

### CODE REVIEW

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 1]
```diff
@@ -452,7 +287,7 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
                         loss = torch.dot(self._criterion_nored(output, target), weights / weights.sum())
                     else:
                         loss = self._criterion(output, target)
-                stopw.stop("Loss")
```
Comment (robinholzi) [inline modyn/trainer_server/internal/trainer/pytorch_trainer.py:455]: I added a level of indentation here, I suppose that's fair?

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py [hunk 2]
```diff
@@ -57,7 +57,7 @@
     seed_everything,
 )
 
-AvailableQueues = Enum("AvailableQueues", ["TRAINING", "DOWNSAMPLING"])
+AvailableQueues = Literal["TRAINING", "DOWNSAMPLING"]
```
Comment (robinholzi) [inline modyn/trainer_server/internal/trainer/pytorch_trainer.py:60]: I think it's appropriate to use literals here as the are more lightweight.

**Comment by XianzheMa** [comment]
The time measurements there are not reliable. See https://github.com/eth-easl/modyn/issues/396

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py (hunk 1 above)
Comment (MaxiBoether) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:456]: Yes, probably it will not make a difference, unless exiting the amp scope for some reason is an expensive operation. Good catch!

**Review by MaxiBoether — APPROVED** [review]
Thank you! I gave my reasoning for using enums, maybe it makes sense to you, maybe it doesn't :D I like enums for avoiding typos / silent failures but I would also be ok with changing this.

File: modyn/trainer_server/internal/trainer/pytorch_trainer.py (hunk 2 above)
Comment (robinholzi) [reply modyn/trainer_server/internal/trainer/pytorch_trainer.py:60]: Fair, probably that's related to flaky linting. In theory, linters / mypy can catch illegal literal values.
